More precise Integer type inference, and fix #51236

Chen768959 opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
More precise Integer type inference, fix #51236


### Documentation entry for user-facing changes

`FieldToDataType<on_error>::operator() (const Array & x)` and `FieldToDataType<on_error>::operator() (const Map & map)` will be able to recognize whether an UInt64 is unsigned, signed, or both, which means that an Int will have multiple possible types. This will make the getLeastSupertype function more accurate.
Example :
select  [-4741124612489978151, -3236599669630092879, 5607475129431807682]
'-4741124612489978151' is Int64, '-3236599669630092879' is Int64, '5607475129431807682' is UInt64 or Int64.
So, their least supertype is Int64

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

Chen768959 committed:
More precise Integer type inference

robot-ch-test-poll4 commented:
<!-- automatic status comment for PR #53003 from Chen768959/ClickHouse:fix-51236 -->
This is an automated comment for commit e797d877480386cb38dcdd5034129b9f4a469ef1 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53003/e797d877480386cb38dcdd5034129b9f4a469ef1/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

Chen768959 committed:
add test.
fix style.

Chen768959 committed:
add switch break

Chen768959 committed:
Add the optimize_type_ids parameter to control whether optimization is enabled.

Chen768959 committed:
fix optimizeTypeIds, Some scenarios do not require optimization:
if UInt64(possible: Int64), UInt64(possible: Int64) = UInt64, UInt64
if Int64(possible: UInt32), Int64(possible: UInt32) = Int64, Int64

Chen768959 committed:
modify comments

Chen768959 committed:
temporarily disable the new Integer type inference feature to verify if it is causing the zk validation failure issue.

Chen768959 committed:
getPossiblePtr no longer has a default value, optimize the optimizeTypeIds method.

Chen768959 committed:
Optimize the code and streamline the functionality by retaining only the feature of automatically recognizing UInt64 as Int64.

Chen768959 committed:
Temporarily modify ErrorCodes to avoid conflicts.

Chen768959 committed:
Merge branch 'master' into fix-51236

Chen768959 committed:
Merge branch 'master' into fix-51236

Chen768959 commented:
@Avogar I have completed all the modifications, and I believe this pull request is now ready for review. I would greatly appreciate it if you could take a look when you have some free time.

Avogar commented:
Sure, I will look at it today

Avogar commented on src/DataTypes/DataTypesNumber.h:
```diff
@@ -9,10 +10,21 @@
 public:
+    DataTypeNumber() = default;
+
+    explicit DataTypeNumber(DataTypePtr opposite_sign_data_type_)
```
Actually, I am not sure it's good to modify general `DataTypeNumber`/`IDataType` class for using it only in one specific place. It just could be confusing why it's needed for future readers. I think it will be better to add a separate class derived from `DataTypeNumber` for this perpose. We can name it like `DataTypeBothSignsNumber` and add a comment why we need it and where it's used (I am not good at names, you can think of better one). And in `getLeastSupertype` instead of calling `hasOppositeSignDataType` we will just try to cast type to `DataTypeBothSignsNumber` (using `typeid_cast`/`dynamic_cast` for example, no need for new methods in `IDataType`).

Avogar commented on src/DataTypes/getLeastSupertype.h:
```diff
@@ -18,7 +18,7 @@ enum class LeastSupertypeOnError
   */
 template <LeastSupertypeOnError on_error = LeastSupertypeOnError::Throw>
-DataTypePtr getLeastSupertype(const DataTypes & types);
+DataTypePtr getLeastSupertype(const DataTypes & types, bool optimize_type_ids = false);
```
I think we don't need the new flag here, let's just do it always, it should not brake anything

Avogar commented on src/DataTypes/getLeastSupertype.cpp:
```diff
@@ -601,6 +603,55 @@ DataTypePtr getLeastSupertype(const DataTypes & types)
     return throwOrReturn<on_error>(types, "", ErrorCodes::NO_COMMON_TYPE);
 }
 
+void optimizeTypeIds(const DataTypes & types, TypeIndexSet & type_ids)
```
Let's write a small comment explaining what this function does

Avogar commented on src/DataTypes/getLeastSupertype.cpp:
```diff
@@ -601,6 +603,55 @@ DataTypePtr getLeastSupertype(const DataTypes & types)
+    // Determine whether UInt64 in type_ids needs to change its sign.
+    for (const auto & type : types)
+    {
+        auto type_id = type->getTypeId();
+
+        if (!has_signed_int)
+            has_signed_int = is_signed_int(type_id);
```
We can use just:
```c++
has_signed_int |= WhichDataType(type).isNativeInt();
```
and remove `is_signed_int` lambda

Avogar replied:
Or even do smth like this before the loop
```
   if (!types.contains(TypeIndex::Int8) && !types.contains(TypeIndex::Int16) && !types.contains(TypeIndex::Int32) && !types.contains(TypeIndex::Int64);
       return;
```

Chen768959 commented:
Thank you for the review and suggestions! I will make the necessary changes promptly.

Chen768959 committed:
Addressed review feedback

Chen768959 commented:
@Avogar Hello, I have created a brand new derived class called `DataTypeNumberWithOpposite.` In the `optimizeTypeIds `, I also attempted to use `typeid_cast `to cast type as `DataTypeNumberWithOpposite`. 

However, I encountered an issue where this type can only be cast as `DataTypeNumberWithOpposite `and cannot be cast as `DataTypeUInt64`. As a result, `DataTypeNumberWithOpposite `fails to pass numerous checks for the `DataTypeUInt64 `. I am unsure if my approach is correct, but it seems that there is a conflict between the derived class `DataTypeNumberWithOpposite `and `typeid_cast<DataTypeUInt64>`.

If it's convenient, I would appreciate hearing your perspective on this matter.

Chen768959 commented on src/DataTypes/FieldToDataType.cpp:
```diff
@@ -36,6 +37,7 @@ DataTypePtr FieldToDataType<on_error>::operator() (const UInt64 & x) const
     if (x <= std::numeric_limits<UInt8>::max()) return std::make_shared<DataTypeUInt8>();
     if (x <= std::numeric_limits<UInt16>::max()) return std::make_shared<DataTypeUInt16>();
     if (x <= std::numeric_limits<UInt32>::max()) return std::make_shared<DataTypeUInt32>();
+    if (x <= std::numeric_limits<Int64>::max()) return std::make_shared<DataTypeUInt64WithOpposite>(std::make_shared<DataTypeInt64>());
```
`DataTypeUInt64WithOpposite` type is returned, but this prevents passing through scenarios that require `typeid_cast<const DataTypeUInt64 *>(type)`, such as the `castBothTypes` method in `FunctionBinaryArithmetic`.

Chen768959 committed:
Perform optimization of LeastSupertype within FieldToDataType, eliminating modifications to DataType.

Chen768959 committed:
Merge branch 'ClickHouse:master' into fix-51236

Chen768959 commented:
@Avogar  Could you please take a look at my issue with automatic merging? I'm encountering the error message 'Required statuses must pass before merging' and I'm unsure how to resolve it. Any assistance would be greatly appreciated. Thank you!

Chen768959 committed:
fix style

Chen768959 committed:
Merge branch 'ClickHouse:master' into fix-51236

Avogar commented:
Sorry for long review, I was on vacation.
I see you changed the approach, so we don't change `getLeastSupertype` anymore. That's a really good Idea, I like it. Let me check the details

Avogar commented on src/DataTypes/FieldToDataType.cpp:
```diff
@@ -136,8 +136,39 @@ DataTypePtr FieldToDataType<on_error>::operator() (const Array & x) const
     DataTypes element_types;
     element_types.reserve(x.size());
 
+    auto checkIfConversionSigned = [&](bool& has_signed_int, bool& has_uint64, bool& uint64_could_opposite,
```
I see we have 2 identical lambdas here and during Map processing. Let's move this code to a separate function in anonimous namespace in this cpp file.

Avogar replied:
And also the logic is the same here and later during Map processing. Maybe we can extract all this logic to a separate function?
Like
`void convertUInt64ToInt64IfPossible(DataTypes & data_types)`

So the code here will look like this:
```
...
for (const Field & elem : x)
        element_types.emplace_back(applyVisitor(*this, elem));
convertUInt64ToInt64IfPossible(element_types);
...
```

It will be much better for future readers

Chen768959 replied:
OK, I'll handle it.

Chen768959 commented:
> Sorry for long review, I was on vacation.

No worries, I hope you had a great vacation!

Chen768959 committed:
Merge branch 'ClickHouse:master' into fix-51236

Chen768959 committed:
extract checkUInt64ToIn64Conversion and convertUInt64ToInt64IfPossible to separate functions

Avogar reviewed (approved):
Looks good, thanks for the contribution!

Avogar commented:
I just realize that this approach with changes only in `FieldToDataType` is actually not general, because it won't work with nested types and some functions like `map/tuple`:
```
:) select [[-4741124612489978151, 1, -3236599669630092879], [2, 5607475129431807682]]

Received exception:
Code: 386. DB::Exception: There is no supertype for types UInt64, Int64 because some of them are signed integers and some are unsigned integers, but there is no signed integer type, that can exactly represent all required unsigned integer values: While processing [[-4741124612489978151, 1, -3236599669630092879], [2, 5607475129431807682]]. (NO_COMMON_TYPE)

:) select map(-4741124612489978151, 1, -3236599669630092879, 2, 5607475129431807682, 3)

Received exception:
Code: 386. DB::Exception: There is no supertype for types UInt64, Int64 because some of them are signed integers and some are unsigned integers, but there is no signed integer type, that can exactly represent all required unsigned integer values: While processing map(-4741124612489978151, 1, -3236599669630092879, 2, 5607475129431807682, 3). (NO_COMMON_TYPE)
```

I think we can get back to the approach with changes in `getLeastSuperType` and separate constructor for `DataTypeNumber` (yes, I didn't like it, but as I remember there were problems with separate type because of `typeid_cast`, so I think it's ok).

Avogar commented:
No worries, I will do such changes by myself, so, no need to do anything, I just wanted you to know.
